base/inference.jl

```diff
@@ -3767,23 +3778,23 @@ function inlineable(f::ANY, ft::ANY, e::Expr, atypes::Vector{Any}, sv::Inference
                                   atype_unlimited, method.sig)
         methsp = methsp::SimpleVector
     end
-    # check whether call can be inlined to just a quoted constant value
-    if isa(f, widenconst(ft)) && !method.isstaged && (method.source.pure || f === return_type)
-        if isconstType(e.typ)
-            return inline_as_constant(e.typ.parameters[1], argexprs, sv,
-                                      invoke_data)
-        elseif isa(e.typ,Const)
-            return inline_as_constant(e.typ.val, argexprs, sv,
-                                      invoke_data)
-        end
-    end
 
     methsig = method.sig
     if !(atype <: metharg)
         return invoke_NF(argexprs, e.typ, atypes, sv, atype_unlimited,
                          invoke_data)
     end
 
+    # check whether call can be inlined to just a quoted constant value
+    if isa(f, widenconst(ft)) && !method.isstaged && (method.source.pure || f === return_type) &&
+        are_args_const(atypes)
```

Comment:
Here's a simple example of the failure scenario – it doesn't even need arguments :P – but it's just a minor rewrite of #20704:
```julia
julia> function method_error end
method_error (generic function with 0 methods)

julia> Base.@pure a() = (method_error(); 1)
a (generic function with 1 method)

julia> b() = a()
b (generic function with 1 method)

julia> b()
1
```

I think we either need to recompute the Const here, or add a bool field to const (`was_computed_by_apply_pure`) that we can check here.